### DISCUSSION TITLE
Simulatneous use of the -json and -debug flags

### DISCUSSION THREAD

**Discussion by savushkin-yauheni** [projectdiscovery/nuclei#1573] [Q&A] [answered]
Hi team

Is there any way to use `json` and `debug` flag simultaneously ?

I would like to pass json option to get only valid results in json.
And would like to provide debug file to log all responses. 

Is it  possible at all ?
Right now as i checked `json` disable `debug` at all. 

Thanks for your time

**Comment by savushkin-yauheni** [comment]
it depends on the order

works. and this version outputs all info to console, no test.json file after.
```
 -debug -json test.json
```

doesn't work:
```
 -json test.json  -debug
```

But anyway. How separate outputs ?

Valid results to json file.
Debug responses to another ?

**Comment by forgedhallpass** [comment, marked as answer]
The `-debug` flag is to display requests and responses on the output. If you'd like to persist them in the json file, you can use `-include-rr` (see `nuclei -help`). If you'd like your requests and responses saved in a separate file, you can use [`proxify`](https://github.com/projectdiscovery/proxify) to achieve that.

**Reply by savushkin-yauheni** [reply]
Hi @forgedhallpass 
Thanks for your answer

1. `include-rr` is a good point, but it's only for findings, not for all requests
2. `proxify` it may work. But don't you think it's an overhead in this case ? 
We have logger already. I'm not familiar with [gologger](https://github.com/projectdiscovery/gologger), but maybe we can just add file handler to `DEBUG` logger and it will work pretty well ?

Because it looks like a task for logger, not for external tool.

**Reply by forgedhallpass** [reply]
> `proxify` it may work. But don't you think it's an overhead in this case ?

I haven't done a benchmark to see the overhead, but I don't expect it to be significant. If you decide to do one yourself, let us know your findings. For your use-case, it might be the best solution though.

> We have logger already. I'm not familiar with [gologger](https://github.com/projectdiscovery/gologger), but maybe we can just add file handler to `DEBUG` logger and it will work pretty well?

It's not that simple. `Gologger` and the whole logging approach will (have to) be reworked in https://github.com/projectdiscovery/nuclei/issues/1166, but even if we include the part of persisting the logs, it'll persist everything within the configured log level (as generally expected from loggers), and you'll have to do the filtering (manually) afterwards. Currently the `debug` flag (not a log level per-se) refers to request and responses only (`-debug` = `-debug-req` + `-debug-resp`), which from a logging perspective is not correct, and will be changed.

> Because it looks like a task for logger, not for external tool.

Depends on what do you mean by external tools :) Our solutions were/are designed to be modular, hence the "separation", but they work together nicely. I assume (haven't checked though) if you would like to see the results of the burp scanner (e.g through `logger++`), it would probably run everything through a proxy.

**Reply by savushkin-yauheni** [reply]
I see. Okay, i will try proxify. Thanks!

Btw, could you check https://github.com/projectdiscovery/nuclei-templates/discussions/3646 ?
I think there is a problem in nuclei itself.

**Reply by savushkin-yauheni** [reply]
Hi again @forgedhallpass 

Is it everything okay with proxify cert ? 

1. I tried on my ubuntu VPS -> `Cannot read TLS response from mitm'd server proxyconnect tcp`

2. Tried on my local Mac.  ->
```
2022/02/10 11:34:36 [241] WARN: Cannot read TLS response from mitm'd server proxyconnect tcp: dial tcp :0: connect: can't assign requested address
```

In both cases I have installed certificate. 
Thanks

**Reply by forgedhallpass** [reply]
Could you please confirm if it's reproducible on the [`dev`](https://github.com/projectdiscovery/proxify/tree/dev) branch?

**Reply by savushkin-yauheni** [reply]
Indeed. Everything  is okay on dev branch. Thanks!
But it's a little bit strange that `Installation` way doesn't work =)

**Reply by forgedhallpass** [reply]
It could have been because of a [bug](https://github.com/projectdiscovery/proxify/pull/82). The next release will contain the fix.
